server-common/src/main/java/org/apache/kafka/queue/KafkaEventQueue.java

```diff
@@ -178,15 +190,18 @@ public void run() {
         }
 
         private void remove(EventContext eventContext) {
-            eventContext.remove();
+            boolean removed = eventContext.remove();
             if (eventContext.deadlineNs.isPresent()) {
-                deadlineMap.remove(eventContext.deadlineNs.getAsLong());
+                removed |= deadlineMap.remove(eventContext.deadlineNs.getAsLong()) != null;
                 eventContext.deadlineNs = OptionalLong.empty();
             }
             if (eventContext.tag != null) {
                 tagToEventContext.remove(eventContext.tag, eventContext);
                 eventContext.tag = null;
             }
+            if (removed) {
+                onQueueSizeChange.accept(size.decrementAndGet());
```

Comment:
I am a bit sceptical about running an arbitrary consumer in a synchronized block of code (this runs by acquiring a lock). Because the caller can potentially perform expensive things in callback which could cause block the other threads from accessing the lock, hence, block critical execution while performing a side-effect. Instead, could we make it an async explicit metric publish event?

Reply:
The goal isn't to have an arbitrary consumer, we just need to propagate the changes to `ControllerMetrics`. I'm not sure what you mean with "an async explicit metric publish event", but I agree it makes senses that we should be careful while holding the lock. I think we can invert the logic and instead of pushing the size change to `ControllerMetrics`, we can use a Supplier instead to pull it.